### ISSUE TITLE
Syncing java targets builds resources.

### ISSUE THREAD

**Issue by uri-canva** [bazelbuild/intellij#1785]
When syncing java targets the class jars are built, which requires building all resources. It's not clear why that would be necessary since the resources don't seem to be used for code indexing.

This causes syncing to take longer than necessary.

This might be due to there not being an output for a jar without resources in https://docs.bazel.build/versions/3.0.0/skylark/lib/java_output.html.

**Comment by uri-canva** [comment]
I have noticed in https://github.com/bazelbuild/intellij/blob/c279bd605818d660a37e5a118c5a30761e821fb5/base/src/com/google/idea/blaze/base/ideinfo/LibraryArtifact.java#L88 we end up not using the class jar at all if a source jar is available. That makes sense: if we don't have a source jar, we can refer to the decompiled class jar, but if we have the source jar we can match that with the interface jar.

It seems wasteful to build the class jar just to have it ignored later on, so I have changed the aspect to not build it if source jars are available. See https://github.com/bazelbuild/intellij/pull/1786.

**Comment by uri-canva** [comment]
ping @brendandouglas

**Comment by brendandouglas** [comment]
Thanks for looking into this.

I have some concerns with the change which I'd need to spend some time investigating before committing to it:

1. is there any guarantee that rules which declare source jar outputs at analysis time actually output them at build time?  Similarly, what if a rule produces no ijars, only class jars and source jars?

2. there are situations where we will decide to drop source jars unconditionally, especially when working from home with additional security restrictions or an average network connection -- in those cases, build time (when building remotely) is not an issue, and we want to minimize the size of the jars we're copying to the local machine. We'd also prefer class jars over interface jars when available in that case. Switching this behavior at the aspect level is much more complex than doing it post-build at the plugin level (output groups are our best means of passing 'flags' to the aspect, and we already have too many of those)

**Comment by brendandouglas** [comment]
There's also the question of how much time we're actually saving by skipping class jars -- do you have some data on this?

**Comment by uri-canva** [comment]
> There's also the question of how much time we're actually saving by skipping class jars -- do you have some data on this?

I don't have hard data, but some of our java targets include closure compiled javascript or rust binaries as resources, building those resources takes over twice as long as building the java code in the whole repo, with those types of compilation being so time consuming.

So this patch is very important for us.

**Comment by uri-canva** [comment]
> is there any guarantee that rules which declare source jar outputs at analysis time actually output them at build time? Similarly, what if a rule produces no ijars, only class jars and source jars?

Not outputting an output declared at analysis time at build time is an error: from https://docs.bazel.build/versions/3.4.0/skylark/rules.html#actions

> **If an action does not generate a file that is listed in its outputs**: This is an execution error and the build will fail. This happens for instance when a compilation fails.

> there are situations where we will decide to drop source jars unconditionally, especially when working from home with additional security restrictions or an average network connection

That's very interesting, I didn't realise that was possible. How do you configure that?

> Switching this behavior at the aspect level is much more complex than doing it post-build at the plugin level (output groups are our best means of passing 'flags' to the aspect, and we already have too many of those)

Now that starlark configurations exist we could use that maybe? https://docs.bazel.build/versions/3.4.0/skylark/config.html

I'm happy to have a look if you point me to the right places in the code.

**Comment by uri-canva** [comment]
I've looked into this more and noticed the logic in `JavaSourceFilter` that determines if something is considered a library, and so would need the class jar, or a source target, in which case it wouldn't, is pretty involved. I have duplicated it in starlark in our internal fork to avoid building  the class jars where not necessary, but I think it's pretty ugly to have to keep two parallel implementations of the same logic in sync across two different languages, so I  don't  think I want to submit it as a PR  here.

In the meantime I have filed a feature request upstream for class jars containing only classes and not resources: https://github.com/bazelbuild/bazel/issues/12195. I believe that would achieve at least a major part of what I'm trying to do by skipping resources in a much less invasive way, and it would still work with the feature you mention here:

> there are situations where we will decide to drop source jars unconditionally, especially when working from home with additional security restrictions or an average network connection
